src/GafferOSL/ShadingEngine.cpp

```diff
@@ -384,6 +471,20 @@ class RenderState
 			return false;
 		}
 
+		std::string applyAttributeSubstitutions( std::string s ) const
+		{
+			if( m_attributeSubstitutions )
+			{
+				return stringApplySubstitutions( s, m_attributeSubstitutions );
+			}
+			else
+			{
+				// I kinda wonder if this should be an error that would make it more obvious that you've
+				// failed to provide attributes?
```

Comment:
If I understand correctly, this would error only when an `<attr:>` substitution is found, and you haven't turned on `OSLObject.useAttributes`? But would still replace with an empty string if you did turn on `OSLObject.useAttributes` and the attribute doesn't exist? That sounds fairly reasonable...

Reply:
Yeah, that was exactly what I was thinking.  Tried implementing this, but I don't see any way to get exceptions to pass through OSL, so I guess I'll leave it as is.